evm/src/cpu/control_flow.rs

```diff
@@ -66,20 +61,30 @@ pub fn eval_packed_generic<P: PackedField>(
     nv: &CpuColumnsView<P>,
     yield_constr: &mut ConstraintConsumer<P>,
 ) {
+    // The null special instruction must be boolean.
+    yield_constr.constraint(lv.null * (lv.null - P::ONES));
+    // Once we reach a null row, there must be only null rows.
+    yield_constr.constraint_transition(lv.null * (nv.null - P::ONES));
+
     let is_cpu_cycle: P = COL_MAP.op.iter().map(|&col_i| lv[col_i]).sum();
     let is_cpu_cycle_next: P = COL_MAP.op.iter().map(|&col_i| nv[col_i]).sum();
-    // Once we start executing instructions, then we continue until the end of the table.
-    yield_constr.constraint_transition(is_cpu_cycle * (is_cpu_cycle_next - P::ONES));
+    // Once we start executing instructions, then we continue until the end of the table
+    // or we reach dummy padding rows.
+    yield_constr
+        .constraint_transition(is_cpu_cycle * (is_cpu_cycle_next - P::ONES) * (nv.null - P::ONES));
 
     // If a row is a CPU cycle and executing a native instruction (implemented as a table row; not
     // microcoded) then the program counter is incremented by 1 to obtain the next row's program
     // counter. Also, the next row has the same kernel flag.
     let is_native_instruction: P = NATIVE_INSTRUCTIONS.iter().map(|&col_i| lv[col_i]).sum();
     yield_constr.constraint_transition(
-        is_native_instruction * (lv.program_counter - nv.program_counter + P::ONES),
+        (nv.null - P::ONES)
+            * is_native_instruction
+            * (lv.program_counter - nv.program_counter + P::ONES),
+    );
+    yield_constr.constraint_transition(
+        (nv.null - P::ONES) * is_native_instruction * (lv.is_kernel_mode - nv.is_kernel_mode),
```

Comment:
I would not change this. That way we can verify that by the time we get to the null rows, execution actually _is_ done. In other words, we need to make sure that the prover can't just start inserting null rows before the entire program has correctly finished.

Reply:
The first constraint would fail if unchanged though, because at the last actual CPU row, the PC won't be increased (as kept at `halt_pc` for all subsequent padding rows).

Reply:
Actually nevermind! As with `PANIC` below the `halt` label we need to adapt slightly the padding to directly write dummy rows when reaching the `halt_pc` the former constraint becomes valid again!